cpp/src/arrow/dataset/file_csv_test.cc

```diff
@@ -130,12 +139,16 @@ TEST_F(TestCsvFileFormat, IsSupported) {
   ASSERT_OK_AND_ASSIGN(supported, format_->IsSupported(*source));
   ASSERT_EQ(supported, false);
 
-  source = GetFileSource();
+  source = GetFileSource(R"(f64
+1.0
+
+N/A
+2)");
   ASSERT_OK_AND_ASSIGN(supported, format_->IsSupported(*source));
   EXPECT_EQ(supported, true);
 }
 
-TEST_F(TestCsvFileFormat, DISABLED_NonMaterializedFieldWithDifferingTypeFromInferred) {
+TEST_F(TestCsvFileFormat, NonMaterializedFieldWithDifferingTypeFromInferred) {
```

Comment:
This should have read "NonProjected" rather than "NonMaterialized", sorry for the confusion and I'll amend. `f64` will be materialized (since we need to filter on it) but is not projected (so it won't appear in scanned batches).

The physical schema of the batch will have the inferred type for each column, I'll add an assertion of this.

The key point of this test case is that we request `f64` be read *as a string*, rather than a double. This is uncommon and only possible due to the textual nature of CSV, but it is not illegal. If the csv reader fails to respect the stated dataset schema this can break filter expressions in surprising ways.

For example it might erroneously consider only the projected schema and revert to inferring the type of a non projected field which would result in a comparison between a double and a string column.